packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,421 @@
+import path from 'path'
+
+import type { Capabilities, Frameworks } from '@wdio/types'
+import type { Browser, MultiRemoteBrowser } from 'webdriverio'
+import { BeforeCommandArgs, AfterCommandArgs } from '@wdio/reporter'
+
+import { v4 as uuidv4 } from 'uuid'
+import type { Pickle, ITestCaseHookParameter } from './cucumber-types'
+
+import { getCloudProvider, getGitMetaData, getHookType, getScenarioExamples, getUniqueIdentifier, getUniqueIdentifierForCucumber, isBrowserstackSession, isScreenshotCommand, removeAnsiColors, uploadEventData } from './util'
+import type { TestData, TestMeta, PlatformMeta, UploadType } from './types'
+
+export default class InsightsHandler {
+
+    private _browser?: Browser<'async'> | MultiRemoteBrowser<'async'>
+    private _tests: { [index: string]: TestMeta }
+    private _hooks: { [index: string]: string[] }
+    private _platformMeta?: PlatformMeta
+    private _framework?: string
+    private _commands: { [index: string]: BeforeCommandArgs & AfterCommandArgs }
+    private _gitConfigPath?: string
+
+    constructor (framework?: string) {
+        this._tests = {}
+        this._hooks = {}
+        this._commands = {}
+        this._framework = framework
+    }
+
+    async setUp (browser: Browser<'async'> | MultiRemoteBrowser<'async'>, browserCaps?: Capabilities.Capabilities, isAppAutomate?: boolean, sessionId?: string) {
```

Comment:
I wonder if it makes more sense to do most of this stuff in the constructor.

Reply:
Correct me if I am wrong, `browser` object we get in `before` hook in service and we don't have that in worker service constructor, right?
If it isn't available in constructor then has to be called in before..
Or we mean that we should move the initialisation of InsightsHandler to before hook itself?